### DISCUSSION TITLE
Attach device to multiple clients simultaneously?

### DISCUSSION THREAD

**Discussion by ramiabughazaleh** [dorssel/usbipd-win#925] [General]
Hi.

Thank you for usbipd-win.

Is it possible to attach a single shared USB device to multiple clients simultaneously?

For example,
1. Install usbipd-win (server) on machine A (Windows Server 2022)
2. Share the device (ex. run `usbipd bind --busid=3-1`)
3. Install usbip (client) on machine B (Windows 10)
4. Attach to the device (ex. run `usbip.exe attach -r machineA -b 3-1`
5. Install usbip (client) on machine C (Windows 10)
6. Attach to the device (ex. run `usbip.exe attach -r machineA -b 3-1`

When I try to do step 6, I get the following error:
`usbip: error: already used bus id: 3-1`
or using a different usbip client:
```
Cannot attach machineA:3240/3-1
Error 0x1020002
Device busy (already exported).
```

Should this be possible?
For example, I'm expecting to share a USB device and single operations to be atomic and exclusive between machines B and C.
If machine B wants to perform a device operation, it will obtain a lock.
If machine C wants to perform a device operation at the same time, it will wait until the operation by machine B is completed by waiting until the lock is released, so it can obtain its own lock on the device.

I'm not very knowledgeable of how USB hardware and how read/write operations may work, and it probably shows with this question. 😇
I found an old thread where others were trying to do the same thing, but no definitive answer:
https://sourceforge.net/p/usbip/discussion/418507/thread/30e03b95/

If software is the only limiting factor, could this feature be added to usbipd-win?
Or is there another solution I can take advantage of?

Thank you.

**Comment by dorssel** [comment]
No, this cannot be done. And in general it cannot be done by USB. A device can be used by one machine/driver at a time. You need higher level software (such as file server, printer server, etc.) to make group access possible. At the device level, the device is used exclusively by one driver only.

**Reply by w2918000** [reply]
Is it possible to create a manager at the driver layer to manage multiple TCP/IP connections and unified access to USB devices? There is no device binding between USBIP-WIN2 and USBIPD-WIN, only TCP/IP connections are made
